File: SVDstart.py

sushkov = False

if sushkov:
    read_movie_file = open('/home/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/data/mu/data1.txt', 'r')
    read_user_file = open('/home/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/data/um/data1.txt', 'r')
    qual = open('/home/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/data/mu/qual.dta', 'r')
    write_file = open('/home/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/submissions/SVDstart.txt', 'w')
else:
    read_movie_file = open('C:\Users\Mike\Documents\Mike\Caltech\\aJunior Fall\CS 156b\mu\data2.txt', 'r')
    read_user_file = open('C:\Users\Mike\Documents\Mike\Caltech\\aJunior Fall\CS 156b\um\data2.txt', 'r')
    qual = open('C:\Users\Mike\Documents\Mike\Caltech\\aJunior Fall\CS 156b\mu\qual.dta', 'r')
    write_file = open('C:\Users\Mike\Documents\Mike\Caltech\\aJunior Fall\CS 156b\k_means.txt', 'w')

##############################################
# HELPER METHODS
##############################################

# trains our algorithm given an input data point
def trainSVD(user, movie, rating):
    numIterations = 0

    # large so we go into the first iteration of the while loop
    error = 10^9
    assert(abs(LEARNING_RATE * (rating - predictRating(movie, user)) - error) > EPSILON)
    
    # learn while our error is decreasing by an amount higher than epsilon
    # or until we hit 200 iterations
    while abs(LEARNING_RATE * (rating - predictRating(movie, user)) - error) > EPSILON and numIterations < 2:
        # calculate the new error
        error = LEARNING_RATE * (rating - predictRating(movie, user))
    
        # iterate over features 0 through 39
        for k in range(NUM_FEATURES):
            oldUserFeature = userFeatures[user][k]
            userFeatures[user][k] += LEARNING_RATE * (error * movieFeatures[movie][k] - K * oldUserFeature)
            movieFeatures[movie][k] += LEARNING_RATE * (error * oldUserFeature - K * movieFeatures[movie][k])
            
        numIterations += 1

# calculates the dot product of the feature vectors for a given user and movie
def predictRating(movie, user):
    return dotProduct(userFeatures[user][:], movieFeatures[movie][:]) 
    
# calculate the dot product of 2 lists
def dotProduct(list1, list2):
    assert(len(list1) == len(list2))
    
    result = 0.0
    for i in range(len(list1)):
        result += (float(list1[i]) * float(list2[i]))
    return result

##############################################
# END HELPER METHODS
##############################################

# constants
NUM_FEATURES = 40
NUM_USERS = 458293
NUM_MOVIES = 17770
LEARNING_RATE = 0.001
K = 25
EPSILON = 10^-3

# initialize feature matrices (2D arrays); initialize to 0.1
# feature k for user u is userFeatures[u][k]
userFeatures = [[0.1] * NUM_FEATURES] * NUM_USERS
movieFeatures = [[0.1] * NUM_FEATURES] * NUM_MOVIES

print 'Starting...'

numLinesSoFar = 0

# go through the input data line by line
for line in read_movie_file:
    # line: user_i movie_j date rating
    vals = line.split()
    
    curr_user = int(vals[0])
    curr_movie = int(vals[1])
    curr_rating = int(vals[3])
    
    # call the training method using the given data point
    trainSVD(curr_user, curr_movie, curr_rating)
    
    numLinesSoFar += 1
    if numLinesSoFar % 1000 == 0:
        print numLinesSoFar

# at this point we have userFeatures and movieFeatures completely populated

# write out the results
for line in qual:
    vals = line.split()
    user = int(vals[0])
    movie = int(vals[1])
    
    # calculate the rating
    predicted_score = calculateScore(user, movie)
    
    # output the data to the submission file
    write_file.write(str(predicted_score) + '\n')

# close the files
write_file.close()
qual.close()    

print 'DONE!'

File: c++/SVDstart.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include <sstream>
#include <cassert>

#include <stdio.h>
#include <stdlib.h>
#include <cmath>

using namespace std;

const int MAC = 0;

// Filenames
const char dataFilePath[] = "/home/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/data/mu/data1.txt";
const char qualFilePath[] = "/home/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/data/mu/qual.dta";
const char outputFilePath[] = "/home/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/submissions/SVDstart.txt";

const char dataFilePathMac[] = "/users/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/data/mu/data2.txt";
const char qualFilePathMac[] = "/users/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/data/mu/qual.dta";
const char outputFilePathMac[] = "/users/msushkov/Dropbox/Caltech/Caltech Classes/CS/CS 156b/submissions/SVDstart.txt";

// constants
const int NUM_FEATURES = 15;
const int NUM_USERS = 458293;
const int NUM_MOVIES = 17770;
const double LEARNING_RATE = 0.001;
const int K = 0.02;
const int EPSILON = 10^-5;
const int MAX_ITERATIONS = 1;
const int NUM_EPOCHS = 80;

// defines a single training data point
class DataPoint {
public:
    int user;
    int movie;
    int rating;

    DataPoint() {
        this->user = 0;
        this->movie = 0;
        this->rating = 0;
    }

    DataPoint(int user, int movie, int rating) {
        this->user = user;
        this->movie = movie;
        this->rating = rating;
    }
};

// a vector of datapoints to hold data
vector<DataPoint*> *trainingData;

// feature matrices (2D arrays)
// feature k for user u is userFeatures[u][k]
double **userFeatures = new double*[NUM_USERS];
double **movieFeatures = new double*[NUM_MOVIES];

// initialize each element of the feature matrices to 0.1
void initialize() {
    // initialize the user feature matrix to all 0.1's
    for (int i = 0; i < NUM_USERS; i++)
        userFeatures[i] = new double[NUM_FEATURES];
    for (int i = 0; i < NUM_USERS; i++) {
        for (int j = 0; j < NUM_FEATURES; j++)
            userFeatures[i][j] = 0.1;
    }
        
    // initialize the movie feature matrix to all 0.1's
    for (int i = 0; i < NUM_MOVIES; i++)
        movieFeatures[i] = new double[NUM_FEATURES];
    for (int i = 0; i < NUM_MOVIES; i++) {
        for (int j = 0; j < NUM_FEATURES; j++)
            movieFeatures[i][j] = 0.1;
    }

    trainingData = new vector<DataPoint*>;
}

// helper function that splits a string
vector<string> &split(const string &s, char delim, vector<string> &elems) {
    stringstream ss(s);
    string item;
    while(getline(ss, item, delim))
        elems.push_back(item);
    return elems;
}

// helper function to split a string
vector<string> split(const string &s, char delim) {
    vector<string> elems;
    return split(s, delim, elems);
}

// calculates the dot product of the feature vectors for a given user and movie
double predictRating(int movie, int user) {
	assert(user < NUM_USERS);
	assert(user >= 0);
	assert(movie >= 0);
	assert(movie < NUM_MOVIES);

    double result = 0.0;
    for (int i = 0; i < NUM_FEATURES; i++)
        result += (userFeatures[user][i]) * (movieFeatures[movie][i]);
    return result;
}

// trains our algorithm given an input data point
void trainSVD(int user, int movie, int rating) {
	assert(user <= NUM_USERS);
	assert(user > 0);
	assert(movie > 0);
	assert(rating > 0);
	assert(rating <= 5);
	assert(movie <= NUM_MOVIES);

    int numIterations = 0;

    // learn while our error is decreasing by an amount higher than epsilon
    // or until we hit some # of iterations
    while (numIterations < MAX_ITERATIONS) {       
        // iterate over features 0 through 39
        for (int k = 0; k < NUM_FEATURES; k++) {
            double error = rating - predictRating(movie - 1, user - 1);

            double oldUserFeature = userFeatures[user - 1][k];
            userFeatures[user - 1][k] += LEARNING_RATE * (error * movieFeatures[movie - 1][k] - K * oldUserFeature);
            movieFeatures[movie - 1][k] += LEARNING_RATE * (error * oldUserFeature - K * movieFeatures[movie - 1][k]);
        }
            
        numIterations++;
    }
}

// iterate through the epochs and all the data points for each epoch
void learn() {
    // go N times through the data
    for (int i = 0; i < NUM_EPOCHS; i++) {
        cout << "Current epoch: " << i + 1 << " out of " << NUM_EPOCHS << endl;

        // go through each point in the data set
        for (int p = 0; p < trainingData->size(); p++) {
            // train the SVD on each point (will go through all features)
            trainSVD(trainingData->at(p)->user, trainingData->at(p)->movie, trainingData->at(p)->rating);
        }
    }
}

// reads data in
void getData() {
    string line;
    ifstream dataFile;
    
    if (MAC)
    	dataFile.open(dataFilePathMac, ios::in);
    else
    	dataFile.open(dataFilePath, ios::in);
    
    int numLinesSoFar = 0;

    // go through the input data line by line
    while (getline(dataFile, line)) {
        // where are we in the current line?
        int count = 0;
        DataPoint *currPoint = new DataPoint();

        istringstream lineIn(line);
        while (lineIn) {
            int val = 0;
            if (lineIn >> val) {
                if (count == 0)
                    currPoint->user = val;
                else if (count == 1)
                    currPoint->movie = val;
                else if (count == 3)
                    currPoint->rating = val;
                count++;
            }
        }
        // add the point to our data vector
        trainingData->push_back(currPoint);
            
        numLinesSoFar++;
        if (numLinesSoFar % 1000000 == 0)
            cout << numLinesSoFar / 1000000 << " million" << endl;
    }    
    dataFile.close();
}

// writes data out
void outputResults() {
    // output file
    ofstream outputFile;
    if (MAC)
    	outputFile.open(outputFilePathMac, ios::out);
    else
    	outputFile.open(outputFilePath, ios::out);
    
    // qual file
    string line;
    ifstream qualFile;
    if (MAC)
    	qualFile.open(qualFilePathMac, ios::in);
    else
    	qualFile.open(qualFilePath, ios::in);
    
    // go through each line of the qual file
    while (getline(qualFile, line)) {
        // where are we in the current line?
        int count = 0;

        int user = -1;
        int movie = -1;

        // read the values on the current line one by one
        istringstream lineIn(line);
        while (lineIn) {
            int val = 0;
            if (lineIn >> val) {
                if (count == 0)
                    user = val;
                else if (count == 1)
                    movie = val;
                count++;
            }
        }

        // calculate the rating for user, movie
        double predictedScore = predictRating(movie - 1, user - 1);

        // write to file
        outputFile << predictedScore << endl;
    }

    qualFile.close();
    outputFile.close();
}

void cleanup() {
    for (int i = 0; i < NUM_USERS; i++)
        delete[] userFeatures[i];
        
    for (int i = 0; i < NUM_MOVIES; i++)
        delete[] movieFeatures[i];
        
    delete[] userFeatures;
    delete[] movieFeatures;

    for (int i = 0; i < trainingData->size(); i++)
        delete trainingData->at(i);
    delete trainingData;
}

int main() {
    initialize();

    getData();
    cout << "Done with getData()" << endl;
    
    // run the SVD algorithm
    learn();
    cout << "Done learning" << endl;

    outputResults();
    cout << "Done with outputResults()" << endl;

    cleanup();
    cout << "DONE!" << endl;
    
    return 0;
}